test/state/mpt.cpp

```diff
@@ -143,108 +119,83 @@ void MPTNode::insert(const Path& path, bytes&& value)  // NOLINT(misc-no-recursi
     // - if m_path is shorter and matches the path prefix
-    //   then insert_idx points at path[m_path.size()].
-    assert(insert_idx != path.end() && "a key must not be a prefix of another key");
-
-    const Path common{m_path.begin(), this_idx};
-    const Path insert_tail{insert_idx + 1, path.end()};
+    //   then insert_idx_it points at path[m_path.size()].
+    assert(insert_idx_it != path.end() && "a key must not be a prefix of another key");
+    const Path insert_tail{insert_idx_it + 1, path.end()};
 
-    switch (m_kind)
+    if (m_kind == Kind::branch && this_idx_it == m_path.end())  // Paths match: go into the child.
     {
-    case Kind::branch:
-    {
-        assert(m_path.empty());  // Branch has no path.
-        if (auto& child = m_children[*insert_idx]; child)
-            child->insert(insert_tail, std::move(value));
+        if (auto& child = m_children[*insert_idx_it]; child)
+            child->insert(insert_tail, std::move(value));  // ①
         else
-            child = leaf(insert_tail, std::move(value));
-        break;
-    }
-
-    case Kind::ext:
-    {
-        assert(!m_path.empty());       // Ext must have non-empty path.
-        if (this_idx == m_path.end())  // Paths match: go into the child.
-            return m_children[0]->insert({insert_idx, path.end()}, std::move(value));
-
-        // The original branch node must be pushed down, possible extended with
-        // the adjusted extended node if the path split point is not directly at the branch node.
-        // Clang Analyzer bug: https://github.com/llvm/llvm-project/issues/47814
-        // NOLINTNEXTLINE(clang-analyzer-cplusplus.NewDeleteLeaks)
-        auto this_branch = optional_ext({this_idx + 1, m_path.end()}, std::move(m_children[0]));
-        auto new_leaf = leaf(insert_tail, std::move(value));
-        *this =
-            ext_branch(common, *this_idx, std::move(this_branch), *insert_idx, std::move(new_leaf));
-        break;
+            child = leaf(insert_tail, std::move(value));  // ②
     }
-
-    case Kind::leaf:
+    else  // ③: Shorten path of this node and insert it to the new branch node.
     {
-        assert(!m_path.empty());  // Leaf must have non-empty path.
-        assert(this_idx != m_path.end() && "a key must be unique");
-        auto this_leaf = leaf({this_idx + 1, m_path.end()}, std::move(m_value));
-        auto new_leaf = leaf(insert_tail, std::move(value));
-        *this =
-            ext_branch(common, *this_idx, std::move(this_leaf), *insert_idx, std::move(new_leaf));
-        break;
-    }
-
-    default:
-        assert(false);
+        const Path extended_path{m_path.begin(), this_idx_it};
+        const auto this_idx = *this_idx_it;
+        m_path = Path{this_idx_it + 1, m_path.end()};  // shorten this path, invalidates this_idx_it
+        *this = branch(extended_path, this_idx, std::make_unique<MPTNode>(std::move(*this)),
+            *insert_idx_it, leaf(insert_tail, std::move(value)));
```

Comment:
What if we have a method to copy node with children, and try to simplify like:
```suggestion
        // The original node must be pushed down.
        const Path common_path{m_path.begin(), this_idx_it};
        const Path this_node_tail{this_idx_it + 1, m_path.end()};
        auto this_node = copy_node(m_kind, this_node_tail, std::move(m_children), std::move(m_value)); // this_idx_it is not invalidated
        auto new_leaf = leaf(insert_tail, std::move(value));
        
        *this = branch(common_path, *this_idx_it, std::move(this_node), *insert_idx_it, std::move(new_leaf));
```

Reply:
I don't think we need special method. The move constructor can do this. I believe I tried this but the effect was more code similarly confusing. I'm still thinking about some refactoring here (e.g. removing named constructors `leaf()` and `branch()`) but wanted to minimize the diff this time.